src/Autocomplete/src/Doctrine/EntityMetadata.php

```diff
@@ -42,21 +42,65 @@ public function isEmbeddedClassProperty(string $propertyName): bool
     }
 
     public function getPropertyMetadata(string $propertyName): array
+    {
+        trigger_deprecation('symfony/ux-autocomplete', '2.15.0', 'Calling EntityMetadata::getPropertyMetadata() is deprecated. You should stop using it, as it will be removed in the future.');
+
+        try {
+            return $this->getFieldMetadata($propertyName);
+        } catch (\InvalidArgumentException $e) {
+            return $this->getAssociationMetadata($propertyName);
+        }
+    }
+
+    /**
+     * @internal
+     *
+     * @return array<string, mixed>
+     */
+    public function getFieldMetadata(string $propertyName): array
     {
         if (\array_key_exists($propertyName, $this->metadata->fieldMappings)) {
-            return $this->metadata->fieldMappings[$propertyName];
+            $metadata = $this->metadata->fieldMappings[$propertyName];
+
+            // In doctrine/orm:^3.0; $metadata will be a FieldMapping object
+            if (!\is_array($metadata)) {
+                return (array) $metadata;
+            }
+
+            return $metadata;
         }
 
+        throw new \InvalidArgumentException(sprintf('The "%s" field does not exist in the "%s" entity.', $propertyName, $this->metadata->getName()));
+    }
+
+    /**
+     * @internal
+     *
+     * @return array<string, mixed>
+     */
+    public function getAssociationMetadata(string $propertyName): array
+    {
         if (\array_key_exists($propertyName, $this->metadata->associationMappings)) {
-            return $this->metadata->associationMappings[$propertyName];
+            $metadata = $this->metadata->associationMappings[$propertyName];
+
+            // In doctrine/orm:^3.0; $metadata will be an AssociationMapping object
+            if (!\is_array($metadata)) {
+                return (array) $metadata;
+            }
+
+            return $metadata;
         }
 
         throw new \InvalidArgumentException(sprintf('The "%s" field does not exist in the "%s" entity.', $propertyName, $this->metadata->getName()));
     }
 
     public function getPropertyDataType(string $propertyName): string
     {
-        return $this->getPropertyMetadata($propertyName)['type'];
+        try {
+            return $this->getFieldMetadata($propertyName)['type'];
+        } catch (\InvalidArgumentException $e) {
```

Comment:
Could we avoid handling exception ? Is there a way to detect which method to use ?